### PR TITLE
Implement flash messages

### PR DESCRIPTION
**Pull request by jonocodes**
This creates the 'flash' function which can be called from any view. Then the next page load will display the message.

It also added messages to the 'create', 'edit', 'delete' functions.

Screenshot from 2024-05-15 13-02-22

### CODE REVIEW

**Comment by hasansezertasan** [comment]
Looks like a quite good implementation.

It's similar to `litestar`'s flash plugin and `starlette-flash` package implementation.

Using session to keep the flash messages in the flash bag on redirects is a great idea but the user could use a custom `SessionMiddleware` that depends on a key value databade like Redis which might cause conflicts. I believe instead of using sessions to store flash messages, we can directly use cookies with a customizable cookie name default to `admin_flash_messages` to store everything.

Or we can come up with a middleware that is specifically for flashes, something like `FlashMiddleware`.

File: sqladmin/helpers.py [hunk 1]
```diff
@@ -26,6 +27,13 @@
 T = TypeVar("T")
 
 
+class AlertTypeEnum(enum.Enum):
+    success = "success"
+    info = "info"
+    warning = "warning"
+    danger = "danger"
```
Comment (hasansezertasan) [inline sqladmin/helpers.py:34]: Why limit the message type? I think this could be just a string.

**Review by hasansezertasan — CHANGES_REQUESTED** [review]
Looks like a quite good implementation. Great job 👍.

It's similar to `litestar`'s flash plugin and `starlette-flash` package implementation.

Using sessions to keep the flash messages in the flash bag on redirects is a great idea but the user could use a custom `SessionMiddleware` that depends on a key-value database like Redis might cause conflicts. I believe instead of using sessions to store flash messages we can directly use cookies with a customizable cookie name default to `admin_flash_messages` to store everything.

Or we can come up with a middleware that is specifically for flashes, something like `FlashMiddleware` 🤓

Also I do think we should update `edit.html` and `create.html` to use flash bag instead of plain error key check.

File: sqladmin/helpers.py (hunk 1 above)
Comment (jonocodes) [reply sqladmin/helpers.py:34]: These map to tabler alert types https://tabler.io/docs/components/alerts

Comment (hasansezertasan) [reply sqladmin/helpers.py:34]: That's a good point but I believe that this shouldn't be limited by an enumeration.

Even though flash messages have some basic categories, they could have custom categories.

WDYT @aminalaee?

Comment (jonocodes) [reply sqladmin/helpers.py:34]: Looks like django uses an enum https://docs.djangoproject.com/en/5.0/ref/contrib/messages/#adding-a-message

**Comment by jonocodes** [comment]
> Using sessions to keep the flash messages in the flash bag on redirects is a great idea but the user could use a custom `SessionMiddleware` that depends on a key-value database like Redis might cause conflicts. I believe instead of using sessions to store flash messages we can directly use cookies with a customizable cookie name default to `admin_flash_messages` to store everything.

Yeah we can do this in cookies, it you guys think its better.

> Also I do think we should update `edit.html` and `create.html` to use flash bag instead of using plain `error` variable.

Yes.

File: sqladmin/helpers.py (hunk 1 above)
Comment (hasansezertasan) [reply sqladmin/helpers.py:34]: Oh I see...

**Comment by zbyte64** [comment]
I think this may have been an instance of perfection being the enemy of good.

Requiring this to be done with cookies changes the contract in many places that only expose the request object. Consider the auth interface, how difficult it would be for it to flash a message that their password will expire if it has to manipulate the response object to do so.

What's the unhappy path here if their session backend isn't compatible? Can we give them an alternative so that the vast majority of deployments isn't blocked on this feature?

**Pull request closed without merging**